src/buffer/out/ImageSlice.cpp

```diff
@@ -0,0 +1,235 @@
+    {
+        auto dstIterator = MutablePixels(dstWriteBegin, dstWriteEnd);
+        auto srcIterator = srcSlice.Pixels(srcUsedBegin);
+        const auto writeCellCount = dstWriteEnd - dstWriteBegin;
+        const auto writeByteCount = sizeof(RGBQUAD) * writeCellCount * _cellSize.width;
+        for (auto y = 0; y < _cellSize.height; y++)
+        {
+            std::memmove(dstIterator, srcIterator, writeByteCount);
+            std::advance(srcIterator, srcSlice._pixelWidth);
+            std::advance(dstIterator, _pixelWidth);
+        }
+    }
+
+    // The used destination before and after the written area must be erased.
+    if (dstUsedBegin < dstWriteBegin)
+    {
+        _eraseCells(dstUsedBegin, dstWriteBegin);
+    }
+    if (dstUsedEnd > dstWriteEnd)
+    {
+        _eraseCells(dstWriteEnd, dstUsedEnd);
+    }
+
+    // If the beginning column is now not less than the end, that means the
+    // content has been entirely erased, so we return true to let the caller
+    // know that the slice should be deleted.
+    return _columnBegin >= _columnEnd;
+}
+
+void ImageSlice::EraseBlock(TextBuffer& buffer, const til::rect rect)
+{
+    for (auto y = rect.top; y < rect.bottom; y++)
+    {
+        auto& row = buffer.GetMutableRowByOffset(y);
+        EraseCells(row, rect.left, rect.right);
+    }
+}
+
+void ImageSlice::EraseCells(ROW& row, const til::CoordType columnBegin, const til::CoordType columnEnd)
+{
+    auto& imageSlice = row.GetMutableImageSlice();
+    if (imageSlice) [[unlikely]]
+    {
+        const auto scale = row.GetLineRendition() != LineRendition::SingleWidth ? 1 : 0;
+        if (imageSlice->_eraseCells(columnBegin << scale, columnEnd << scale))
+        {
+            // If _eraseCells returns true, that means the image was
+            // completely erased, so we can delete this slice.
+            imageSlice = nullptr;
+        }
+    }
+}
+
+bool ImageSlice::_eraseCells(const til::CoordType columnBegin, const til::CoordType columnEnd)
+{
+    if (columnBegin <= _columnBegin && columnEnd >= _columnEnd)
+    {
+        // If we're erasing the entire range that's in use, we return true to
+        // indicate that there is now nothing left. We don't bother altering
+        // the buffer because the caller is now expected to delete this slice.
+        return true;
+    }
+    else
+    {
+        const auto eraseBegin = std::max(columnBegin, _columnBegin);
+        const auto eraseEnd = std::min(columnEnd, _columnEnd);
+        if (eraseBegin < eraseEnd)
+        {
+            const auto eraseOffset = (eraseBegin - _columnBegin) * _cellSize.width;
+            const auto eraseLength = (eraseEnd - eraseBegin) * _cellSize.width;
+            auto eraseIterator = _pixelBuffer.begin() + eraseOffset;
+            for (auto y = 0; y < _cellSize.height; y++)
+            {
+                std::fill_n(eraseIterator, eraseLength, RGBQUAD{});
+                std::advance(eraseIterator, _pixelWidth);
+            }

```

Comment:
The for loop advances the iterator by `_cellSize.height * _pixelWidth` in total which is equal to the `_pixelBuffer` size. If the `eraseOffset` is greater than 0, the final loop iteration will leave the iterator past the end of the `_pixelBuffer`. Since the MSVC STL uses checked iterators this results in a debug assertion.

One way to fix this:
```cpp
auto eraseIterator = _pixelBuffer.begin();
for (auto y = 0; y < _cellSize.height; y++)
{
    std::fill_n(eraseIterator + eraseOffset, eraseLength, RGBQUAD{});
    std::advance(eraseIterator, _pixelWidth);
}
```

FYI `fill_n` isn't super duper optimal for clearing bytes in this loop and you may be better off using `memset` directly. Mostly because it skips an unnecessary emptiness check. Something like this:
```cpp
auto eraseIterator = _pixelBuffer.data() + eraseOffset;
for (auto y = 0; y < _cellSize.height; y++)
{
    memset(eraseIterator, 0, eraseLength * sizeof(RGBQUAD));
    eraseIterator += _pixelWidth;
}
```

Reply:
Thanks for bringing this up. It never occurred to me that it would be a problem using `std::advance` passed the end of the buffer, but it makes perfect sense in retrospect. I think I only starting using it because I was getting complaints from the auditor about pointer arithmetic in some places, so I'll be happy to go back to `+=`. If there are still pointer arithmetic warnings I'll find another way to deal with that.

Reply:
Turns out `+=` isn't any better than `std::advance` since it still has the debug asserts. But I found that switching to raw pointers - and continuing to use the `std::advance` for incrementing - was enough to keep everyone happy. Debug build doesn't appear to assert anymore, and audit still passes.